### PR TITLE
Add AOT gzip content-encoding support for main build files.

### LINKED ISSUE #555: Ahead-of-time GZIP compression
Upon `next build`. This is for the next major.
- [x] For main build files - https://github.com/zeit/next.js/pull/565
- [x] For JSON pages - https://github.com/zeit/next.js/pull/571
- [ ] Allow to extend compression via `next.config.js`- https://github.com/zeit/next.js/pull/657

### PR DESCRIPTION
**Pull request by arunoda**
Related to #555 

Currently, we only do this for `main.js` and `commons.js` only.
We can do this for individual `.json` pages as well.

For that, we may need to change a bit of how we save pages and so on.
I think we could handle it in a different PR.

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -212,6 +214,22 @@ export default class Server {
...
+    }
+
+    const gzipPath = `${path}.gz`
+    const exists = await fs.exists(gzipPath)
+    if (!exists) {
+      return await this.serveStatic(req, res, path)
+    }
+
+    res.setHeader('Content-Encoding', 'gzip')
+    return await this.serveStatic(req, res, gzipPath)
```
Comment (rauchg) [inline server/index.js]: `return await` is not needed. can just be `return`

**Comment by rauchg** [comment]
Would it be easy to write a test?

**Comment by rauchg** [comment]
Adjusted the ticket to divide the task in two parts per your recommendation @arunoda

File: server/index.js (hunk 1 above)
Comment (arunoda) [inline server/index.js]: Done.

**Comment by arunoda** [comment]
@rauchg Yep.
I not sure about to have a test case since we don't do a production build right now.
May be we need to have a different test setup, targeting individual examples.

File: server/build/gzip.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import fs from 'fs'
+import path from 'path'
+import zlib from 'zlib'
+
+export default async function gzipAssets (dir) {
+  const nextDir = path.resolve(dir, '.next')
+
+  await gzip(path.resolve(nextDir, 'commons.js'))
+  await gzip(path.resolve(nextDir, 'main.js'))
```
Comment (rauchg) [inline server/build/gzip.js]: `await Promise.all` could have a perf advantage here

Comment (arunoda) [inline server/build/gzip.js]: Yeah. Parallelism.

Comment (arunoda) [inline server/build/gzip.js]: Done.

**Pull request merged**